Distribution logic

dvush opened the pull request:
## 📝 Summary

This implements a redistribution logic for the builder. It adds redistribution logic as part of the backtest and adds a binary to calculate output csv files.

## 💡 Motivation and Context

Redistribution refers to a process of dividing block profits between contribution signers.

To do that we
* built block and calculate its backtest profit
* exclude some signer and calculate profit as if this signer never sent any bundles
* the difference between two values (second one should be lower) is contribution by the given signer
* profit is divided proportionally to the contribution



For transactions sent through the rpc endpoint we can't use signer and instead we use tx origin or address specified as address to receive mev share refunds.

---

## ✅ I have completed the following steps:

* [ ] Run `make lint`
* [ ] Run `make test`
* [ ] Added tests (if applicable)

dvush committed:
initial

dvush committed:
use coinbase signer for baktest

dvush committed:
protect txs

dvush committed:
add filter address data

dvush committed:
filter out mempool txs

dvush committed:
rebase fixes

ZanCorDX commented on crates/rbuilder/src/backtest/redistribute/cli/csv_output.rs:
```diff
@@ -0,0 +1,52 @@
+        Ok(Self { file })
+    }
+
+    pub fn write_header(&mut self) -> io::Result<()> {
```
Can write_header() be called inside new() and make it private?

ZanCorDX commented on crates/rbuilder/src/backtest/redistribute/cli/mod.rs:
```diff
@@ -0,0 +1,134 @@
+        HistoricalDataStorage::new_from_path(&config.base_config.backtest_fetch_output_file)
+            .await?;
+
+    match cli.command {
```
Can we avoid the copy/paste on Block vs Range?

ZanCorDX commented:
Mempool txs are brought from Duration::minutes(3) ago (MempoolDumpsterDatasource) but in live buiding spawn_clean_orderpool_job keeps only TIME_TO_KEEP_BUNDLE_CANCELLATIONS: Duration = Duration::from_secs(60); are we handling this?

ferranbt commented:
I understand that the refund logic needs to be open source for the TEE builder and that part of that refund logic takes advantage of the bench tooling of the Rbuilder, but, could we move this logic to some TEE-builder specific repository?

This logic is part of a product offering from Flashbots and I do not think it belongs into the OSS version of the rbuilder.

ZanCorDX reviewed (changes requested):
Mentioned copy/paste and timing stuff.

ryanschneider commented on crates/rbuilder/src/backtest/backtest_build_range.rs:
```diff
@@ -399,6 +402,7 @@ impl CSVResultWriter {
 fn spawn_block_fetcher(
     mut historical_data_storage: HistoricalDataStorage,
     blocks: Vec<u64>,
+    ingored_signers: Vec<Address>,
```
Typo: 
```suggestion
    ignored_signers: Vec<Address>,
```

ryanschneider commented on crates/rbuilder/src/backtest/backtest_build_range.rs:
```diff
@@ -408,13 +412,16 @@ fn spawn_block_fetcher(
                 }
             };
+            for block in &mut blocks {
+                block.filter_out_ignored_signers(&ingored_signers);
```
```suggestion
                block.filter_out_ignored_signers(&ignored_signers);
```

dvush commented:
I found some issues here that needs to be addressed.

@ferranbt I think the core logic should be in this repo but maybe some wrapper can be in another. The motivation is to have this updated with the rest of the repo and ease of maintenance.

ferranbt commented:
> The motivation is to have this updated with the rest of the repo and ease of maintenance.

What is the maintenance burden? The bench framework?

MoeMahhouk committed:
feat: Add the ability to read the cl_node_urls from env variables (#59)

## 📝 Summary

This PR adds the ability to parse the cl_node_url passed from an
environment variable too.
It would help with the automatic deployment in the TDX yocto images.

This addition does not affect the original behaviour of accepting a
single value or an array of values. It adds extra a way to pass the
values from an env variable where the values are separated by a comma
from a single string.

## 💡 Motivation and Context

The motivation behind this change was triggered when preparing the TDX
demo with rbuilder and the rbuilder was crashing because we were passing
the CL_NODE_URL as environment variables during deployment from the
cloud init conf.
This lead us to hardcode it first just for the sake of the demo but I
believe it would be better to allow this to be passed and fetched from
an environment parameter too similar to the secrets configs.


## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [x] Added tests (if applicable)

---------

Co-authored-by: ZanCorDX <126988525+ZanCorDX@users.noreply.github.com>
Co-authored-by: jn <nxqd.inbox+rhacker@gmail.com>
Co-authored-by: Ferran Borreguero <ferran.borreguero@gmail.com>
Co-authored-by: Ryan Schneider <ryanleeschneider@gmail.com>
Co-authored-by: Alejo Salles <alejosalles@gmail.com>
Co-authored-by: Vitaly Drogan <vitalii.drohan@gmail.com>

chirag-bgh committed:
chore(deps): bump reth to v1.0.3 (#51)

## 📝 Summary

<!--- A general summary of your changes -->

Update reth to latest v1.0.3 #47 

## 💡 Motivation and Context

<!--- (Optional) Why is this change required? What problem does it
solve? Remove this section if not applicable. -->

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [ ] Run `make test`
* [ ] Added tests (if applicable)

zzzzzzch committed:
fix: delete unused features of alloy-chains (#66)

## 📝 Summary

<!--- A general summary of your changes -->
hi, this PR 
+ fixes #65 compilation warnings for unused features of `alloy-chains`
+ ~~fixes the lint errors of `doc_lazy_continuation`, which introduced
in Rust 1.80~~

<!--- (Optional) Why is this change required? What problem does it
solve? Remove this section if not applicable. -->

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [ ] Added tests (if applicable)

dvush committed:
blob gas used

dvush committed:
Merge branch 'distribute' into develop-merge-distribute

dvush committed:
typo

dvush committed:
simplify code

dvush commented:
Diff is super wierd, its actually rebased. Closing and reopening
